### ISSUE TITLE
get CI\CD pipeline working

### ISSUE THREAD

**Issue by jschalk** [jschalk/keg#70]
Hi Femi! Thanks for joining the party,

This project is very complex and has been built using Test Driven Development (https://www.youtube.com/watch?v=yfP_v6qCdcs&t=2s) Maintaining the tests is very time-consuming but it allows for major refactoring with confidence. There are been many many major refactorings and this project would not be possible without TDD. 
Testing tool is only PyTest. Jaar has been built around PyTest's simplest tools. 
Goals:
- [x] have a pipeline that requires all tests to pass before every commit
Sometimes the hardest work is creating a test that identifies the bug. The fix is actually easy after that. 
- [x] identify way to commit failing tests with a flag or something that identifies "good test, fix to come"

**Comment by FemiWilliams** [comment]
Notes/references to self:

CI/CD First Pass
Run tests on commit(push):
https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#push

Building and testing python: https://docs.github.com/en/actions/automating-builds-and-tests/building-and-testing-python

Packaging python projects: https://packaging.python.org/en/latest/tutorials/packaging-projects/

**Comment by FemiWilliams** [comment]
Hey @jschalk,

I created the initial pipeline, that can be viewed on the branch related to this issue. It's set to run the full suite of tests on a push (commits are local to your computer so they can't trigger a github action afaik). I just wanted to get clarification on how you imagine the workflow behaving:

- By 'requires all tests to pass', are you saying that if a test fails, the code should not be merged into the master? If so: 

1. What should happen when one more more tests fail, beyond blocking a merger? How would you like to be notified of a failure? I thought about having the workflow automatically open a ticket with information on the failing tests, but if there's another way you'd like to get that information let me know what works for you.
2. Are you okay with switching to a branching style to make full use of this workflow? i.e. if we have code just being pushed to main, the code will already be on master before the tests run, and we would have to roll back to a previous push or make another push to fix any issues. Having the branches means the tests can fail without affecting the master branch's code, and we can just delete the branch if there's an issue with it.

- For the failing tests, are they tests that we know will fail before they're even pushed? If so, we could have the test commented out and then only 'reveal' it when we expect there to be a fix introduced to the code. Just wanted to get some insight into your thinking here.

**Comment by jschalk** [comment]
Hi Femi, 

Thanks for doing this! And I appreciate very much the clear questions. 

0. Thanks for setting up the pipeline. Lets go forward with your framework: tests run on push 
1. _are you saying that if a test fails, the code should not be merged into the master?_ Yes
2. On push failure we can have it open a ticket as you suggested. If that becomes too cumbersome we'll change it. 
3. _Are you okay with switching to a branching style to make full use of this workflow?_ Yes, I'm willing to change based on your recommendation. In my development experience I've never used branching or a rollback successfully so I've got a blind spot there. I've always fixed the issue and committed to mstr. Please consider these arguments: https://www.youtube.com/watch?v=lXQEi1O5IOI. What you think?
4. _For the failing tests, are they tests that we know will fail before they're even pushed? If so, we could have the test commented out and then only 'reveal' it when we expect there to be a fix introduced to the code. Just wanted to get some insight into your thinking here._ Ok I'll do that. I'll research pytest some more on my own for my edification.

**Comment by jschalk** [comment]
@FemiWilliams My last commit to main: https://github.com/jschalk/jaar/commit/f94963de480b05e3c1667385fa6a8029954a2e58 has 1125 tests that all pass. What next steps should be take do get the pipeline going? Please feel free to assign me work
- [x] assign @jschalk things to do if he can help

**Comment by FemiWilliams** [comment]
1. I took a look at that video and I pretty much agree with the arguments made there, so I want to make the distinction that I'm not suggesting we move to a feature branching approach, but rather we're using the branching function as part of the CI workflow. I went down a bit of a rabbit hole with that youtube channel so I don't remember if it's mentioned in the video you linked or a different one, but he says that with CD we want the code to always be in a releasable state, so this workflow is designed to keep failing code from reaching Master. Since you're writing all the code yourself there shouldn't be an issue, but if more people come on board in the future the focus should be on making tasks (and associated branches) as small as possible to avoid code divergence and merge conflicts rather than associating branches with entire features.
2. Once the workflow is merged into Master it will be a part of all branches created moving forward and should run automatically. I'll go ahead and start that process as it is now and test that things work as expected, and then I'll modify it to automatically delete branches when there are no issues.
3. One thing I forgot to ask in the earlier comment: Do you care about storing the full test results? I was starting to build that into the pipeline but if you're only interested in the details of the failures so they can be fixed rather than holding onto test results I'll focus on other things.
4. I think I'm a little behind that commit, I only see about 1010 tests and 4 are currently failing. Could you take a look at that and see if it's a matter of not having the correct packages or if there's something else I'm missing? If you [go here](https://github.com/jschalk/jaar/actions/runs/8309544974/job/22741026411) and expand 'Execute Tests and Generate Report' you can view the test results.

**Comment by jschalk** [comment]
1. Thanks for going down that rabbit hole. https://www.youtube.com/@ContinuousDelivery) is probably my number one reference for coding practices so if we both know it helps our language be precise. For example I feel like I understand every word you wrote about using the branching function as part of the CI workflow and I agree. Funny comment: when you wrote ..._this workflow is designed to keep failing code from reaching Master. Since you're writing all the code yourself there shouldn't be an issue_ Ha! My code is trash all the time!,
2.  _...I'll go ahead and start that process..._ Great, let me know if we need to significantly restructure the directories. 
3. _Do you care about storing the full test results?_ No, I don't know what they would be useful for. Interesting idea though.
4. The test results show that your branch is at least one major refactor behind. #101 is really changing things. But there is no reason to assume a merge conflict because the external packages are all the same.

**Comment by jschalk** [comment]
- [x] @jschalk create #107 branch
- [x] @jschalk move local env branch to #107

**Comment by jschalk** [comment]
@FemiWilliams: I'm using branches now. Let's move this process to main. Successfully completing this ticket means that on every commit to main every test passes. That's enough for now.

**Comment by FemiWilliams** [comment]
Ok, sounds good. I'm running the tests to make sure everything passes before pushing back to main, but I'm getting an import conflict error between ` /jaar/src/agenda/report.py` and `/jaar/src/world/report.py`. When you test are you running the tests do you have arguments or configurations that you're using to avoid this conflict?

**Comment by jschalk** [comment]
> Ok, sounds good. I'm running the tests to make sure everything passes before pushing back to main, but I'm getting an import conflict error between ` /jaar/src/agenda/report.py` and `/jaar/src/world/report.py`. When you test are you running the tests do you have arguments or configurations that you're using to avoid this conflict?

No I'm not passing anything to avoid conflicts. I checked out your branch and ran the tests locally and did not get any errors. 

I think I made a mistake naming both of them report.py. 
- [ ] ~~in 70-get-cicd-pipeline-working branch change report.py file names. See if testing still works.~~
- [x] in 52-create-main-app-that-builds-personal-agenda-using-scripts-part2 change report.py file names. See if testing still works. 
- [x] push 52 branch changes to main.

**Comment by jschalk** [comment]
I saw that there is no report.py in 70 branch. This brings up an issue with #70 merge to main: the main codebase has moved on from a lot of concepts that are still present in 70 branch. If 70 branch is merged with main there will be multiple deprecated features present. So merging straight up is probably wrong. 

Is it reasonable to do the following?
- [x] create a new issue/branch off main called "get test-on-push working"
- [x] add \jaar\run-tests-on-push.yml to get test-on-push working branch.
- [x] see if github actions work correctly. 
- [ ] merge "get test-on-push working" into main
- [ ] if pushing to main automatically runs the tests close issue "get test-on-push working"
- [ ] delete 70-get-cicd-pipeline-working branch so there is no confusion
- [ ] recreate new 70-get-cicd-pipeline-working branch for any further work in this issue/branch.

**Comment by FemiWilliams** [comment]
Yes that's reasonable. The only change introduced by this branch is the .yml file so it should be fine.

**Comment by FemiWilliams** [comment]
Also minor note to avoid any confusion, it should be .github/workflows/run-tests-on-push.yml

**Comment by jschalk** [comment]
You own this process, do you want me to create the new issue/branch and so forth?

**Comment by FemiWilliams** [comment]
That's true. I'll get it going then and let you know if I run into any issues.

**Comment by jschalk** [comment]
@FemiWilliams I created a branch off main #120 to learn about githubs actions and I saw that issue's branch from main had 7 errors. Some bugs, some OS related. I fixed/changed all tests so they now pass.  I commited the test changed to the main branch and closed #120. I accidently committed the  .github actions yaml to main but then removed it with another pull request.

**Comment by FemiWilliams** [comment]
Sounds good. I'm working on the logs being properly captured so they can be put into an issue when opened, and I'll push it when it's ready.

**Comment by jschalk** [comment]
main branch now runs all tests and creates issue if any test fails. It does not automatically merge.

**Comment by jschalk** [comment]
@FemiWilliams I'm closing this ticket. Please let me know if you want to do more with it.
